File: problems/211/solution.java
class WordDictionary {

    /** Initialize your data structure here. */
    Map <Integer, HashSet<String>> map=new HashMap(); 
    public WordDictionary() {
       
    }
    
    /** Adds a word into the data structure. */
    public void addWord(String word) {
        int l=word.length();
        if(map.containsKey(l)){
            map.get(l).add(word);
        }
        else{
            
            HashSet<String> set=new HashSet();
            set.add(word);
            map.put(l,set);
        }
    }
    
    /** Returns if the word is in the data structure. A word could contain the dot character '.' to represent any one letter. */
    public boolean search(String word) {
        int l=word.length();
        if(!map.containsKey(l)){
            return false;
        }
        else{
            HashSet<String> set = map.get(l);
            int countnoDot=0;  //counter to store numbers of characters with no dot(.)
            int same=0;  //counter to store number of same characters
            for(String s:set){
                countnoDot=0;
                same=0;
                for(int i=0;i<l;i++){
                    if(word.charAt(i)!='.'){
                        countnoDot++;
                        if(word.charAt(i)==s.charAt(i)){
                            same++;
                        }
                    }
                }
                if(same==countnoDot){
                    return true;
                }
            }
            return false;
        }
        
    }
}

File: problems/211/solution.cpp
class TrieNode {
public:
    unordered_map<char, TrieNode*> children;
    bool isWord = false;
    
    TrieNode() {}
};

class WordDictionary {
private:
    TrieNode *root = new TrieNode();
    
    void insert(const string& s, TrieNode *node) {
        TrieNode *cur = node;
        
        for (const char& ch: s) {
            if (not cur->children.count(ch)) {
                cur->children[ch] = new TrieNode();
            }
            
            cur = cur->children[ch];
        }
        
        cur->isWord = true;
    }
    
    bool search(const string& s, TrieNode *node, int pos = 0) {
        TrieNode *cur = node;
        
        for (int i = pos; i < size(s); i++) {
            if (s[i] != '.') {
                if (not cur->children.count(s[i])) {
                    return false;
                }
                
                cur = cur->children[s[i]];
            } else {
                for (auto [ignore, child]: cur->children) {
                    if (search(s, child, i + 1)) {
                        return true;
                    }
                }
                
                return false;
            }
        }
        
        return cur and cur->isWord;
    }
public:
    /** Initialize your data structure here. */
    WordDictionary() {}
    
    void addWord(string word) {
        insert(word, root);
    }
    
    bool search(string word) {
        return search(word, root);
    }
};

/**
 * Your WordDictionary object will be instantiated and called as such:
 * WordDictionary* obj = new WordDictionary();
 * obj->addWord(word);
 * bool param_2 = obj->search(word);
 */